### DISCUSSION TITLE
AttributeError: 'Table' object has no attribute 'email'

### DISCUSSION THREAD

**Discussion by rbracco** [fastapi-users/fastapi-users#915] [Q&A] [answered]
## Describe the bug

Adding fastapi-users to an existing project, with code from a past project that is successfully using fastapi-users results in `AttributeError: 'Table' object has no attribute 'email'` when making a request to the `register` route. This is probably a result of an idiosyncrasy of my code, but I noticed where the issue is occurring in `SQLAlchemyUserDatabase.get_by_email` there was a change in how select statements are done between versions 9.2.1 and 9.2.4 the code changes from `self.users.c.email` to `self.user_table.email` as seen below.  
```python
async def get_by_email(self, email: str) -> Optional[UD]:
        query = self.users.select().where(
            func.lower(self.users.c.email) == func.lower(email)
        )
        user = await self.database.fetch_one(query)
        return await self._make_user(user) if user else None
```
to
```python
async def get_by_email(self, email: str) -> Optional[UD]:
        statement = select(self.user_table).where(func.lower(self.user_table.email) == func.lower(email))
        return await self._get_user(statement)
```

## Expected behavior

Consistent behavior across minor versions, but I recognize that this is probably due to some weird way I've set up my tables.

## Configuration
 - Python version : 3.8.10
 - FastAPI version : 0.73.0
 - FastAPI Users version : 9.2.4

### FastAPI Users configuration
My UserTable is set up in models/user.py as follows. 
![image](https://user-images.githubusercontent.com/47190785/153248222-28b89228-bf2b-4396-bb26-edc249e6f87c.png)

and this is what database.py looks like, note that I have to pass the user table in a weird way to avoid a circular import
![image](https://user-images.githubusercontent.com/47190785/153248398-e27da25e-5b98-4011-a6a5-f43d6928c285.png)

full stack trace:
![image](https://user-images.githubusercontent.com/47190785/153248497-6239f1de-7780-4e0a-9fb5-386c9df09231.png)

## Additional context

I have verified that the users table is successfully being created and has an email field, and that the code is otherwise identical to my other project. Finally changing `english_toolkit.db.models.user.users` to `english_toolkit.db.models.user.UserTable` in database.py gets further but still fails with 
![image](https://user-images.githubusercontent.com/47190785/153250248-ab24a5cd-58d3-4926-b1fc-eda43b27c314.png)

**Comment by frankie567** [comment, marked as answer]
Hi Robert 😉 

I think the problem comes from a sub-dependency mismatch.

The setup you show uses `encode/databases`, which was the way adopted in the first version of SQLAlchemy adapter (<= v1.1.0). However it seems you installed the v2+ version. You can check this with the command:

```sh
pip freeze | grep fastapi-users-db-sqlalchemy
```

Two solutions then:

1. Migrate the code to be compatible with the new version of the adapter, without `encode/databases`. This is what described in the documentation: https://fastapi-users.github.io/fastapi-users/configuration/databases/sqlalchemy/
2. Stick with the previous one. In this case, make sure the FastAPI Users dependency is like this: `fastapi-users[sqlalchemy]` (**not** `fastapi-users[sqlalchemy2]`)

Cheers!

**Reply by rbracco** [reply]
That did it, you are a genius. Thank you.

**Reply by frankie567** [reply]
Thank you for the coffees, Rob! Greatly appreciated 🙏
